NUAttribute.js

```diff
@@ -74,70 +82,78 @@ export default class NUAttribute extends NUObject {
-                    return attrObj.validateObjectValue(attrValue, attrObj);
+                //if value provided
+                switch (attrObj.attributeType) {
+                    case NUAttribute.ATTR_TYPE_STRING:
+                        return attrObj.validateStringValue(attrValue, attrObj);
+
+                    case NUAttribute.ATTR_TYPE_ENUM:
+                        return attrObj.validateEnumValue(attrValue, attrObj);
+
+                    case NUAttribute.ATTR_TYPE_LIST:
+                        return attrObj.validateListValues(attrValue, attrObj);
+
+                    case NUAttribute.ATTR_TYPE_OBJECT:
+                        return attrObj.validateObjectValue(attrValue, attrObj);
+
+                    case NUAttribute.ATTR_TYPE_INTEGER:
+                    case NUAttribute.ATTR_TYPE_FLOAT:
+                    case NUAttribute.ATTR_TYPE_LONG:
+                    case NUAttribute.ATTR_TYPE_TIMESTAMP:
+                        return attrObj.validateNumberValue(attrValue, attrObj);
+
+                    default:
                 }
             }
         }
         return null;
     }
 
     validateListValues(listValues, attrObj) {
-        for (var i = 0; i < listValues.length; i++) {
-            var listElementValue = listValues[i],
-                dataTypeMismatch = false;        
-            if (attrObj.subType === NUAttribute.ATTR_TYPE_INTEGER || attrObj.subType === NUAttribute.ATTR_TYPE_FLOAT) {
-                dataTypeMismatch = (typeof listElementValue !== 'number');
-            } else if (attrObj.subType !== NUAttribute.ATTR_TYPE_ENUM && typeof listElementValue === 'object') {
+        if (typeof listValues !== 'object') {
+            return dataTypeMismatchError(attrObj, listValues);
+        }
+        for (const listElementValue of listValues) {
+            let err = false;
+            if (typeof listElementValue === 'object') {
+                //validate type list of embedded objects
                 if (attrObj.subType && attrObj.subType.getClassName) {
-                    if (listElementValue && listElementValue.getClassName && listElementValue.getClassName() !== attrObj.subType.getClassName()) {
-                        dataTypeMismatch = true;
+                    if (listElementValue.getClassName && listElementValue.getClassName() !== attrObj.subType.getClassName()) {
+                        return dataTypeMismatchError(attrObj, listElementValue, true)
                     }
                 } else {
                     getLogger().warn(`Invalid subType for attribute: ${attrObj.name}, subType: ${attrObj.subType}`);
                 }
             }
-            if (dataTypeMismatch){
-                return new NUAttributeValidationError(attrObj.localName, attrObj.remoteName,
-                    'Invalid data type',
-                    `Data type should be ${attrObj.subType}, but is ${typeof listElementValue}`);
+
+            switch (attrObj.subType) {
```

Comment:
let's return in the switch!

Reply:
@natabal this validation is done for every single element in the list. we do not want to return if result is null (no error in current element) until all elements have been tested. hence the check in line 146 below.

Reply:
@killanch Fine with me but all I was saying is return in the break if you have error! What am I missing?

Reply:
@natabal understood now, updated code now, could you pls review?

Reply:
Looks good. I approved it already! Please merge once tested.